lib/study_cleanup_tools.rb

```diff
@@ -0,0 +1,172 @@
+# study_cleanup_tools.rb
+#
+# module for ensuring that each test build starts with clean schemas and Terra billing projects
+# methods are safeguarded so that they cannot be run on deployed instances
+# methods can be run against a local instance by passing allow_dev_env: true (but not recommended)
+
+module StudyCleanupTools
+
+  # disallow running an cleanup commands on deployed production & staging hosts
+  DISALLOWED_HOSTS = /singlecell(-staging)?\.broadinstitute/
+
+  # disallow running deletes against any of these projects
+  DISALLOWED_BILLING_PROJECTS = %w(single-cell-portal single-cell-portal-staging)
+
+  # delete all studies and remove associated Terra workspaces & Google buckets
+  # can only be invoked in the test or pentest environment
+  #
+  # * *params*
+  #   - +allow_dev_env+: (Boolean) => allow execution of this method in development environment (default: false)
+  #
+  # * *raises*
+  #   - (RuntimeError) => if command cannot be run for failing a security check (hostname, project, environment, etc)
+  #
+  # * *returns*
+  #   - (Integer) => count of studies that were deleted
+  def self.destroy_all_studies_and_workspaces(allow_dev_env: false)
+    raise_exception_unless_true(:permit_environment?) { permit_environment?(allow_dev_env) }
+    raise_exception_unless_true(:permit_hostname?) { permit_hostname? }
+    raise_exception_unless_true(:confirm_delete_request?) { confirm_delete_request? } if !!(allow_dev_env && Rails.env.development?)
+    study_count = 0
+    Study.all.each do |study|
+      # do not mass delete studies in protected projects
+      if permit_billing_project?(study.firecloud_project)
+        study.destroy_and_remove_workspace
+        study_count += 1
+      end
+    end
+    study_count
+  end
+
+  # delete all Terra workspaces in a project that do not have a study associated with them in the given env/schema
+  # will ignore any workspaces created by different service accounts/users
+  # by default will only run against test/pentest environment during continuous integration build
+  #
+  # * *params*
+  #   - +project_name+: (String) => Name of Terra billing project (defaults to FireCloudClient::PORTAL_NAMESPACE )
+  #   - +allow_dev_env+: (Boolean) => allow execution of this method in development environment (default: false)
+  #
+  # * *raises*
+  #   - (RuntimeError) => if command cannot be run for failing a security check (hostname, project, environment, etc)
+  #
+  # * *returns*
+  #   - (Integer) => count of workspaces that were deleted
+  def self.delete_all_orphaned_workspaces(project_name = FireCloudClient::PORTAL_NAMESPACE, allow_dev_env: false)
+    raise_exception_unless_true(:permit_environment?) { permit_environment?(allow_dev_env) }
```

Comment:
I'd just as soon have these be passed as strings rather than symbols, so it's clearer that they're just used for messaging purposes